File: src/lecture2/lecture2_tf2_cpp/src/subscriber.cpp
#include <ros/ros.h>
#include <tf2_ros/transform_listener.h>

int main(int argc, char *argv[])
{
    ros::init(argc, argv, "subscriber", ros::init_options::AnonymousName);
    ros::NodeHandle n;
    tf2_ros::Buffer buffer;
    tf2_ros::TransformListener tl(buffer);
    ros::Rate rate(30);
    while (ros::ok())
    {
        geometry_msgs::TransformStamped ts;
        try
        {
            ts = buffer.lookupTransform("world", "head", ros::Time(0));
        }
        catch (tf2::TransformException &e)
        {
            ROS_WARN("%s", e.what());
            ros::Duration(1).sleep();
            continue;
        }
        ROS_INFO_STREAM(ts.transform.translation.x << "\t" << ts.transform.translation.y << "\t" << ts.transform.translation.z);
        rate.sleep();
    }
    return 0;
}

File: src/lecture2/lecture2_tf2_py/scripts/subscriber.py
#!/usr/bin/env python3
import rospy
import tf2_ros
    
def subscriber():
    rospy.init_node("subscriber", anonymous=True)
    buf = tf2_ros.Buffer()
    tl = tf2_ros.TransformListener(buf)
    rate = rospy.Rate(30)
    while not rospy.is_shutdown():
        try:
            ts = buf.lookup_transform("world", "head", rospy.Time())
        except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException) as e:
            rospy.logwarn(str(e))
            rospy.sleep(rospy.Duration(1))
            continue
        rospy.loginfo("%f\t%f\t%f", ts.transform.translation.x, ts.transform.translation.y, ts.transform.translation.z)
        rate.sleep()

if __name__ == '__main__':
    subscriber()

